artifacts.go

```diff
@@ -301,205 +294,477 @@ func validateArtifact(c *cli.Context) error {
-	var raw *artifact.Raw = nil
-	var err error = nil
-
-	// read header elements first
-	for _, elem := range artifactHeaderElems {
-		// get element form the artifact
-		if getNext {
-			raw, err = r.ReadRaw()
-			if err != nil {
-				return err
-			}
+func unpackArtifact(name string) (string, error) {
+	f, err := os.Open(name)
+	if err != nil {
+		return "", errors.Wrapf(err, "Can not open: %s", name)
+	}
+	defer f.Close()
+
+	// initialize raw reader and writer
+	aReader := areader.NewReader(f)
+	rootfs := handlers.NewRootfsInstaller()
+
+	tmp, err := ioutil.TempFile("", "mender-artifact")
+	if err != nil {
+		return "", err
+	}
+	defer tmp.Close()
+
+	rootfs.InstallHandler = func(r io.Reader, df *handlers.DataFile) error {
+		_, err = io.Copy(tmp, r)
+		return err
+	}
+
+	if err = aReader.RegisterHandler(rootfs); err != nil {
+		return "", errors.Wrap(err, "failed to register install handler")
+	}
+
+	err = aReader.ReadArtifact()
+	if err != nil {
+		return "", err
+	}
+	return tmp.Name(), nil
+}
+
+func repack(from io.Reader, to io.Writer, key []byte,
+	newName string, dataFile string) (*areader.Reader, error) {
+	sDir, err := ioutil.TempDir("", "mender-repack")
+	if err != nil {
+		return nil, err
+	}
+	defer os.RemoveAll(sDir)
+
+	storeScripts := func(r io.Reader, info os.FileInfo) error {
+		sLocation := filepath.Join(sDir, info.Name())
+		f, fileErr := os.OpenFile(sLocation, os.O_WRONLY|os.O_CREATE|os.O_EXCL, 0755)
+		if fileErr != nil {
+			return errors.Wrapf(fileErr,
+				"can not create script file: %v", sLocation)
 		}
-		getNext = true
+		defer f.Close()
 
-		// check if we are not having element out of order
-		if elem.required && !strings.HasPrefix(raw.Name, elem.name) {
-			return errors.Errorf("Invalid artifact, should contain '%s' "+
-				"file , but contains '%s'", elem.name, raw.Name)
-		} else if !elem.required && !strings.HasPrefix(raw.Name, elem.name) {
-			// we have missing optional element; move on to the next one
-			getNext = false
-			continue
+		_, err = io.Copy(f, r)
+		if err != nil {
+			return errors.Wrapf(err,
+				"can not write script file: %v", sLocation)
 		}
+		f.Sync()
+		return nil
+	}
 
-		// check if we are having "manifest" file, which we need to sign
-		if raw.Name == "manifest" {
-			if err = createSignature(raw, w, key); err != nil {
-				return err
-			}
-			continue
+	verify := func(message, sig []byte) error {
+		return nil
+	}
 
-		} else if raw.Name == "manifest.sig" && !force {
-			// we are re-signing the artifact; return error by default
-			return errors.New("Trying to sign already signed artifact; " +
-				"please use force option")
-		} else if raw.Name == "manifest.sig" && force {
-			// just continue here as new signature is already part of tmp artifact
-			continue
+	data := dataFile
+	ar := areader.NewReader(from)
+
+	if dataFile == "" {
+		tmpData, tmpErr := ioutil.TempFile("", "mender-repack")
+		if tmpErr != nil {
+			return nil, tmpErr
 		}
+		defer os.Remove(tmpData.Name())
```

Comment:
Needs `defer tmpData.Close()` after `Remove` (which confusingly will run it first).

Reply:
Fixed!